File: C++/segment_tree.cc
// CopyRight [2015] <Raynald Chung>
#include<iostream>
#include<algorithm>
#include<cstdio>
#include<climits>
#define Maxn 10000

using std::cout;
using std::cin;

class SegmentTree {
 private:
        struct TreeNode {
            int left, right, cover;
            int max, min;
        };
        TreeNode tree[Maxn];

 public:
        SegmentTree() { }

        void Build(int node_num, int left_branch, int right_branch) {
            Tree[node_num].left = left_branch;
            Tree[node_num].right = right_branch;
            if (right_branch - left_branch >= 1) {
                const int mid = (left_branch + right_branch) >> 1;
                tree[node_num].left = left_branch;
                Build(2 * node_num + 1, left_branch, mid);
                tree[node_num].right = right_branch;
                Build(2 * node_num + 2, mid + 1, right_branch);
                tree[node_num].max = max(Tree[2*node_num+1].max, Tree[2*node_num+2].max);
                tree[node_num].min = min(Tree[2*node_num+1].min, Tree[2*node_num+2].min);
            } else {
                tree[node_num].left = tree[node_num].right = left_branch;
                tree[node_num].max = val[left_branch];
                tree[node_num].min = val[left_branch];
            }
        }

        void Print(int node_num) {
            cout << Tree[node_num].left << " " << Tree[node_num].right <<  " "
                << Tree[node_num].max << " " << Tree[node_num].min << endl;
            if (Tree[node_num].right > Tree[node_num].left) {
                Print(2*node_num+1);
                Print(2*node_num+2);
            }
        }

        void Query(int node_num, int left_branch, int right_branch) {
            if (Tree[num].left == left_branch && Tree[num].right == right_branch) {
                maxx = max(Tree[num].max, maxx);
                minx = min(Tree[num].min, minx);
                return;
            }
            const int mid = (tree[num].left + tree[num].rigbt)/2;
            if (left > mid) {
                Query(2 * node_num + 2, left_branch, right_branch);
            } else if (right <= mid) {
                Query(2 * node_num + 1, left_branch, right_branch);
            } else {
                Query(2 * node_num + 1, left_branch, mid);
                Query(2 * node_num + 2, mid + 1, right_branch);
            }
        }
};

int main() {
    int n, m;
    int a, b;

    SegmentTree s;
    cin >> n >> m;
    for (int i = 0; i < n; i++) {
        cin >> val[i];
    }
    s.build(0, 0, n-1);
    // s.print(0);
    for (int i = 0; i < m; i++) {
        cin >> a >> b;
        a--;
        b--;
        maxx = 0;
        minx = INT_MAX;
        s.query(0, a, b);
        printf("%d\n", maxx - minx);
    }
    return 0;
}

File: Kotlin/SegmentTree.kt
class SegmentTree(size: Int) {
    internal class TreeNode {
        var left: Int = 0
        var right: Int = 0
        var cover: Int = 0
        var value: Int = 0
    }

    private val tree: Array<TreeNode> = Array(2 * size) { TreeNode() }

    init { }

    fun build(node_num: Int, left_branch: Int, right_branch: Int) {
        tree[node_num].left = left_branch
        tree[node_num].right = right_branch
        if (right_branch - left_branch >= 1) {
            val mid = (left_branch + right_branch) / 2
            build(2 * node_num + 1, left_branch, mid)
            build(2 * node_num + 2, mid + 1, right_branch)
        }
    }

    fun pushDown(num: Int) {
        if (tree[num].cover > 0) {
            tree[2 * num + 1].cover += tree[num].cover
            tree[2 * num + 2].cover += tree[num].cover
            val mid = (tree[num].left + tree[num].right).shr(1)
            tree[2 * num + 1].value += tree[num].cover * (mid - tree[num].left + 1)
            tree[2 * num + 2].value += tree[num].cover * (tree[num].right - mid)
            tree[num].cover = 0
        }
    }

    fun add(num: Int, left_branch: Int, right_branch: Int, d: Int) {
        if (tree[num].left == left_branch && tree[num].right == right_branch) {
            tree[num].cover += d
            tree[num].value = d * (tree[num].right - tree[num].left + 1)
        } else {
            pushDown(num)
            val mid = (tree[num].left + tree[num].right) / 2
            if (left_branch > mid) {
                add(2 * num + 2, left_branch, right_branch, d)
            } else if (right_branch <= mid) {
                add(2 * num + 1, left_branch, right_branch, d)
            } else {
                add(2 * num + 1, left_branch, mid, d)
                add(2 * num + 2, mid + 1, right_branch, d)
            }
            tree[num].value = tree[num * 2 + 1].value + tree[num * 2 + 2].value
        }
    }

    fun query(num: Int, left_branch: Int, right_branch: Int): Int {
        if (tree[num].left == left_branch && tree[num].right == right_branch) {
            return tree[num].cover
        }
        pushDown(num)
        val mid = (tree[num].left + tree[num].right) / 2
        if (left_branch > mid) {
            return query(2 * num + 2, left_branch, right_branch)
        } else if (right_branch <= mid) {
            return query(2 * num + 1, left_branch, right_branch)
        } else {
            return query(2 * num + 1, left_branch, mid)+ query(2 * num + 2, mid + 1, right_branch)
        }
    }
}